### DISCUSSION TITLE
Database backups in CP are unpleasant

### DISCUSSION THREAD

**Discussion by ccchapman** [craftcms/cms#13610] [Ideas] [answered]
Database backups through the Craft CP present some weak points.

- No meaningful progress indicator
- Slow on large sites (more than one minute is not unusual for us)

We find that database backups are the slowest requests on our Craft sites. This means we adjust our infrastructure timeouts accordingly. However, it would be better if we did not need to set the timeouts to such high numbers to accommodate one outlier.

I think it would also be better as a user if I got notified when the job is done, rather than being expected to wait and come back to the page.

What could be a better solution here? My idea is the database backup is queued and user is emailed once ready.

**Comment by piotrpog** [comment]
I also found out that sometimes database backup just fails on specific servers. I suspect that it might have something to do with db files being packed to zip, maybe these servers did not had support for generating zip files?

**Comment by angrybrad** [comment, marked as answer]
@ccchapman 

> I think it would also be better as a user if I got notified when the job is done, rather than being expected to wait and come back to the page.

That’s a possibility.

> What could be a better solution here?

Not to rely on large database backups that are processed through PHP and a web server. :)

You can keep bumping resources and timeouts, but at some point, backing up a 50GB database and then compressing it through PHP isn’t a good idea.

FTR, Craft is calling out to `mysqldump`/`pg_dump` on the underlying server to process these backups. It has a reasonable set of default flags:

https://github.com/craftcms/cms/blob/develop/src/db/mysql/Schema.php#L153-L228

https://github.com/craftcms/cms/blob/develop/src/db/pgsql/Schema.php#L117-L143

But you can completely override the default commands via https://craftcms.com/docs/4.x/config/general.html#backupcommand to change the behavior.

@piotrpog when it fails, the underlying error should be in either Craft or your web server's error logs.

**Reply by ccchapman** [reply]
> Not to rely on large database backups that are processed through PHP and a web server. :)

I am not sure entirely what this is suggesting; however, we do not rely on these backups for any critical use case. It is just that we (or less technical users in the CMS) expect it to work better than it does now.

> You can keep bumping resources and timeouts, but at some point, backing up a 50GB database and then compressing it through PHP isn’t a good idea.

Our largest DB is 1.3GB uncompressed and `actions/utilities/db-backup` has a 1.15 min TTFB if that helps.

I understand _why_ it is slow, and think it is a prime candidate to have run as a background task.

> But you can completely override the default commands via https://craftcms.com/docs/4.x/config/general.html#backupcommand to change the behavior.

We do this so here is our config for the record.

```
/**
 * Shell command that Craft should execute to create a database backup
 *
 * See: https://github.com/craftcms/cms/issues/12937
 */
->backupCommand(
    "PGPASSWORD=\"{password}\"" .
        " pg_dump" .
        " --dbname={database}" .
        " --host={server}" .
        " --port={port}" .
        " --username={user}" .
        " --if-exists" .
        " --clean" .
        " --no-owner" .
        " --no-privileges" .
        " --no-acl" .
        " --file=\"{file}\"" .
        " --schema={schema}" .
        " --exclude-table-data '{schema}.assetindexdata'" .
        " --exclude-table-data '{schema}.imagetransformindex'" .
        " --exclude-table-data '{schema}.resourcepaths'" .
        " --exclude-table-data '{schema}.sessions'" .
        " --exclude-table-data '{schema}.cache'"
)
```

**Reply by angrybrad** [reply]
> I understand why it is slow, and think it is a prime candidate to have run as a background task.

That's fair - we'd need to build somewhere in the control panel (likely a utility) to download/manage these dump if we push them off to a queue.

> We do this so here is our config for the record.

It might be worth playing with the `--jobs` flag to parallelize the backup process. It’ll hit your server resources harder but (in theory) will speed up the whole process, too.

https://www.postgresql.org/docs/current/app-pgdump.html

**Reply by ccchapman** [reply]
Thanks for this suggestion! Is this known to work with Craft? I tried it out and I got the exception "Could not create backup: the backup file doesn't exist." `pg_dump` needs you to use `--format=directory` with `--jobs=n`, meaning it does not output a single plain-text SQL file as it does by default. From what I can tell looking at Craft's `UtilitiesController@actionDbBackupPerformAction`, it is expecting a SQL file output.

**Reply by angrybrad** [reply]
@ccchapman Ahh... I’ve used it outside of Craft, but not from the control panel. It sounds like we’d need to add some logic for when it's supplied and to account for the difference in output.
